jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/S3Fields.kt

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package software.aws.toolkits.jetbrains.services.s3
+
+import software.amazon.awssdk.services.s3.S3Client
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request
+import java.time.Instant
+
+/**
+ * S3 Key class represents a base class for S3 Directory and S3 Objects
+ */
+
+sealed class S3Key(val bucket: String, val key: String) {
+    val name: String = if (key.endsWith("/")) {
+        key.dropLast(1)
+    } else {
+        key.substringAfterLast("/")
+    }
+}
+
+class S3Bucket(
+    bucket: String,
+    val client: S3Client,
+    val creationDate: Instant
+) : S3Directory(bucket, "", client)
+
+class S3Object(
+    bucket: String,
+    key: String,
+    val eTag: String,
+    val size: Long,
+    val lastModified: Instant,
+    val client: S3Client
+) : S3Key(bucket, key)
+
+open class S3Directory(
+    bucket: String,
+    key: String,
+    private val client: S3Client
+) : S3Key(bucket, key) {
+
+    fun children(): List<S3Key> {
+        val request = ListObjectsV2Request.builder()
+            .bucket(bucket)
+            .delimiter("/")
+            .prefix(key)
+            .build()
+        val response = client.listObjectsV2(request)
+
+        val folders = (response?.commonPrefixes() ?: emptyList())
```

Comment:
instead of the outer wrapper do:

```
response.commonPrefixes()?.map { S3Directory(bucket, it.prefix(), client) } ?: emptyList()
```

Just makes the intent a little clearer